src/coreclr/jit/morph.cpp

```diff
@@ -9555,6 +9555,37 @@ GenTree* Compiler::fgMorphLeaf(GenTree* tree)
     }
     else if (tree->gtOper == GT_FTN_ADDR)
     {
+        // A function pointer address is being used. If this is for a Delegate scenario
+        // we don't need to prepare the method. However, if the function pointer is being used in
+        // an unsafe manner, as a raw function pointer, the the signature should be walked.
+        // The below logic is attempting to see if the function pointer is being passed
+        // to a Delegate .ctor.
```

Comment:
This seems potentially fragile since you're not connecting up the function pointer with its use. So in a complex tree it seems like we could deduce preparation isn't needed when it is.

So you should also check that the user of the `GT_FTN_ADDR` is an arg to the call. You can do this by walking the call arguments.

Reply:
How does this work?

If I do the simple and naive thing for a delegate:
```cs
private static Func<int> Problem()
{
    static int Get() => 1;

    return Get;
}
```
Then by morph I see we have inlined the `.ctor` and have multiple statements around:
```scala
*************** In fgMarkAddressExposedLocals()
LocalAddressVisitor visiting statement:
STMT00000 ( 0x000[E-] ... 0x00C )
               [000005] -AC---------              *  ASG       ref   
               [000004] D------N----              +--*  LCL_VAR   ref    V01 tmp1         
               [000003] --C---------              \--*  CALL help ref    HELPER.CORINFO_HELP_NEWSFAST
               [000002] H----------- arg0            \--*  CNS_INT(h) long   0xd1ffab1e token

LocalAddressVisitor visiting statement:
STMT00003 ( INL01 @ 0x000[E-] ... ??? ) <- INLRT @ ???
               [000013] -A-XG-------              *  ASG       ref   
               [000012] ---XG--N----              +--*  FIELD     ref    _target
               [000006] ------------              |  \--*  LCL_VAR   ref    V01 tmp1         
               [000011] ------------              \--*  LCL_VAR   ref    V01 tmp1         

LocalAddressVisitor visiting statement:
STMT00004 ( INL01 @ 0x007[E-] ... ??? ) <- INLRT @ ???
               [000017] -A-XG-------              *  ASG       long  
               [000016] ---XG--N----              +--*  FIELD     long   _methodPtr
               [000014] ------------              |  \--*  LCL_VAR   ref    V01 tmp1         
               [000015] H-----------              \--*  CNS_INT(h) long   0xd1ffab1e ftn

LocalAddressVisitor visiting statement:
STMT00005 ( INL01 @ 0x00E[E-] ... ??? ) <- INLRT @ ???
               [000021] -A-XG-------              *  ASG       long  
               [000020] ---XG--N----              +--*  FIELD     long   _methodPtrAux
               [000018] ------------              |  \--*  LCL_VAR   ref    V01 tmp1         
               [000001] ------------              \--*  FTN_ADDR  long   RyuJitReproduction.Program.<Problem>g__Get|11_0


LocalAddressVisitor visiting statement:
STMT00002 ( 0x00C[--] ... ??? )
               [000010] ------------              *  RETURN    ref   
               [000009] ------------              \--*  LCL_VAR   ref    V01 tmp1         
```
Since this seems like an optimization (in other words, walking the signature is always correct, but we want to avoid it if possible), it seems like we could mark the `GT_FTN_ADDR` node with some flag, in, say, `fgOptimizeDelegateConstructor` and then look at the flag here.

Reply:
> it seems like we could mark the GT_FTN_ADDR node with some flag, in, say, fgOptimizeDelegateConstructor and then look at the flag here.

That is an interesting idea. @AndyAyersMS ?

Reply:
Yes, this is probably a better way to go about it. `fgOptimizeDelegateConstructor ` already finds the target function node and checks for `GT_FTN_ADDR`.

Reply:
@AndyAyersMS I added a new `GenTreeFlags` value &ndash; `GTF_DELEGATE_TGT`?

Reply:
Hmm, I don't think adding a new "shared" flag value is needed (and it would require some review to make sure there is enough space available, and it won't quite work as a "shared" one right now anyway, since the common mask wasn't adjusted). Can we not make this `GT_FTN_ADDR`-only?

Reply:
>since the common mask wasn't adjusted

That was a mistake. Fixed.

> Can we not make this GT_FTN_ADDR-only?

I can, but that isn't the only case where it is a target. I'd prefer to have it on the `GenTree` since it seems there are cases where having a complex tree as a target is possible. I can add this somewhere else on `GenTree` but it isn't obvious where.

Reply:
> I can add this somewhere else on GenTree but it isn't obvious where.

Right, the only place where this can be put is in the flags.

> I can, but that isn't the only case where it is a target. I'd prefer to have it on the GenTree since it seems there are cases where having a complex tree as a target is possible.

That is true, but it is also true that we only _need_ this information for `GT_FTN_ADDR`, it seems. Slots in the shared flags are somewhat high in value, in general, and in maintenance cost. Concretely: only the `IAT_VALUE` constant address case in `fgMorphLeaf` will preserve the flag today. Is that bug, do we need to preserve the flag? When do we need to preserve the flag? What breaks if we don't? The answers are "no", "we don't" and "nothing", but they need to be found first, with a non-shared flag they would simply not exist.

That aside, `cTreeFlags` also needs to be updated to display the flag.

Reply:
Moved to `GenTreeFptrVal`.